lib/ticks-to-tree.js

```diff
@@ -89,75 +99,183 @@ function ticksToTree (ticks, options = {}) {
               if (callerMatch) S += 2
             }
           }
+          if (sourceMaps !== null) {
+            jsLibraries[key.split(' ')[1]] = true
+          }
         }
         if (kind === 'Unopt') S += 1
         else if (kind === 'Opt') S += 2
       }
-
       if (type && type !== 'JS') name += ' [' + type + (kind ? ':' + kind : '') + ']'
-      return { S, name, value: 0, top: 0 }
+      return { S, name, type, value: 0, top: 0 }
     })
-
     labelInitFrames(stack, regExp)
 
-    addToMergedTree(stack.map(({ S, name, value, top }) => ({ S, name, value, top })))
+    addToMergedTree(trees, stack.map(({ S, name, type, value, top }) => ({ S, name, type, value, top })))
     // mutate original (save another loop over stack + extra objects)
-    addToUnmergedTree(stack)
+    addToUnmergedTree(trees, stack)
   })
 
-  function addToMergedTree (stack) {
-    var lastFrame = null
-    stack.forEach((frame, ix) => {
-      // TODO need an elegant way to place inlined frames into *other* stacks
-      // where the function has disappeared – for now we just remove inlined
-      // stacks from merged graphs
-      if (frame.S > 2) return // skip inlined
-      if (ix > 0) lastFrame.children = lastFrame.children || []
-      const children = (ix === 0) ? merged.children : lastFrame.children
-      const child = children.find(({ name }) => name === frame.name)
-
-      if (child === undefined) children.push(frame)
-      else frame = child
-
-      if (ix === stack.length - 1) frame.top++
-      if (ix === 0) merged.value += 1
-      frame.value++
-
-      lastFrame = frame
+  if (sourceMaps !== null) {
+    await Promise.all(Object.keys(jsLibraries).map(async (jsLibrary) => {
+      return loadSourceMap(jsLibrary, sourceMaps)
+    }))
+  }
+  if (sourceMaps !== null || relativePath) {
+    Object.keys(trees).map((tree) => {
+      trees[tree].children.map((frame, index) => {
+        renameFrame(frame)
+      })
     })
   }
+  Object.keys(loadedSourceMaps).forEach((file) => {
+    if (loadedSourceMaps[file] !== false) {
+      loadedSourceMaps[file].destroy()
+    }
+  })
+
+  return trees
+}
+
+const loadedSourceMaps = {}
+async function loadSourceMap (fileName, sourceMaps) {
+  if (typeof loadedSourceMaps[fileName] === 'undefined') {
+    const sourceMapFiles = [
+      sourceMaps[fileName],
+      sourceMaps[basename(fileName)],
+      fileName + '.map' // dist.js -> dist.js.map
+    ]
+
+    if (existsSync(fileName)) {
+      const lastLine = await readLastLines.read(fileName, 1)
```

Comment:
is there ever the possibility of situation where the last line is blank and the second to last line is the sourceMapUrl? Or do all tools always ensure that the sourceMapUrl is the final line, with no blank line?

Reply:
The [rfc](https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit?pli=1#) states: `The generated code may include a line at the end of the source, with the following form:`. I read this as sourceMappingURL _must be_ the last line, but its a small change to read the last 2 lines instead to accomodate for an accidental line-break.